frontend/src/store/modules/sheet.ts

```diff
@@ -0,0 +1,269 @@
+import axios from "axios";
+import { isEmpty } from "lodash-es";
+
+import * as types from "../mutation-types";
+import { makeActions } from "../actions";
+import type {
+  Sheet,
+  SheetState,
+  CreateSheetState,
+  Principal,
+  ResourceObject,
+  ConnectionContext,
+  TabInfo,
+} from "../../types";
+import { getPrincipalFromIncludedList } from "./principal";
+
+function convertSheet(
+  sheet: ResourceObject,
+  includedList: ResourceObject[]
+): Sheet {
+  return {
+    ...(sheet.attributes as Omit<Sheet, "id" | "creator" | "updater">),
+    creator: getPrincipalFromIncludedList(
+      sheet.relationships!.creator.data,
+      includedList
+    ) as Principal,
+    updater: getPrincipalFromIncludedList(
+      sheet.relationships!.updater.data,
+      includedList
+    ) as Principal,
+    id: parseInt(sheet.id),
+  };
+}
+
+const state: () => SheetState = () => {
+  return {
+    sheetList: [],
+    sheetById: new Map(),
+    sharedSheet: null,
+    isFetchingSheet: false,
+  };
+};
+
+const getters = {
+  currentSheet: (
+    state: SheetState,
+    getters: any,
+    rootState: any,
+    rootGetters: any
+  ) => {
+    const currentTab = rootGetters["tab/currentTab"];
+
+    if (!currentTab || isEmpty(currentTab)) return null;
+
+    const sheetId = currentTab.sheetId;
+
+    return state.sheetById.get(sheetId);
+  },
+  isCreator: (
+    state: SheetState,
+    getters: any,
+    rootState: any,
+    rootGetters: any
+  ) => {
+    const currentUser = rootGetters["auth/currentUser"]();
+    const currentSheet = getters.currentSheet;
+
+    if (!currentSheet) return false;
+
+    return currentUser.id === currentSheet!.creator.id;
+  },
+  accessIsPrivate: (state: SheetState, getters: any) => {
+    return getters?.currentSheet?.visibility === "PRIVATE" ?? false;
+  },
+  accessIsPobject: (state: SheetState, getters: any) => {
+    return getters?.currentSheet?.visibility === "PROJECT" ?? false;
+  },
+  accessIsPublic: (state: SheetState, getters: any) => {
+    return getters?.currentSheet?.visibility === "PUBLIC" ?? false;
+  },
+  isReadOnly: (
+    state: SheetState,
+    getters: any,
+    rootState: any,
+    rootGetters: any
+  ) => {
+    const currentUser = rootGetters["auth/currentUser"]();
+    const { currentSheet, accessIsPrivate, accessIsPobject, accessIsPublic } =
+      getters;
+
+    if (!currentSheet) return true;
+    // creator/owner always can edit
+    if (getters.isCreator) return false;
+
+    // if current user is not owner, check the link access level
+    return (
+      accessIsPrivate ||
+      (accessIsPobject && currentUser.role === "DEVELOPER") ||
```

Comment:
As a security measure, It's better to use the reverse logic currentUser.role != OWNER

Otherwise, if we add a new role not expecting to grant write access, and we forget to add here, then we have a bad leak.

Reply:
Also the logic here does not feel right.

The role on the currentUser is the workspace role. But we need to check the project level role here, which we need to fetch the projectMembership.